packages/trie/src/checkpointDb.ts

```diff
@@ -0,0 +1,156 @@
+import { LevelUp } from 'levelup'
+import { DB, BatchDBOp } from './db'
+
+export const ENCODING_OPTS = { keyEncoding: 'binary', valueEncoding: 'binary' }
+
+export type Checkpoint = {
+  // We cannot use a Buffer => Buffer map directly. If you create two Buffers with the same internal value,
+  // then when setting a value on the Map, it actually creates two indices.
+  keyValueMap: Map<string, Buffer | null>
+  root: Buffer
+}
+
+/**
+ * DB is a thin wrapper around the underlying levelup db,
+ * which validates inputs and sets encoding type.
+ */
+export class CheckpointDB extends DB {
+  public checkpoints: Checkpoint[]
+
+  /**
+   * Initialize a DB instance. If `leveldb` is not provided, DB
+   * defaults to an [in-memory store](https://github.com/Level/memdown).
+   * @param leveldb - An abstract-leveldown compliant store
+   */
+  constructor(leveldb?: LevelUp) {
+    super(leveldb)
+    // Roots of trie at the moment of checkpoint
+    this.checkpoints = []
+  }
+
+  /**
+   * Is the DB during a checkpoint phase?
+   */
+  get isCheckpoint() {
+    return this.checkpoints.length > 0
+  }
+
+  /**
+   * Adds a new checkpoint to the stack
+   * @param root
+   */
+  checkpoint(root: Buffer) {
+    this.checkpoints.push({ keyValueMap: new Map<string, Buffer>(), root })
+  }
+
+  /**
+   * Commits the latest checkpoint
+   */
+  async commit() {
+    const { keyValueMap } = this.checkpoints.pop()!
+    if (!this.isCheckpoint) {
+      // This was the final checkpoint, we should now commit and flush everything to disk
+      const batchOp: BatchDBOp[] = []
+      keyValueMap.forEach(function (value, key) {
+        if (value === null) {
+          batchOp.push({
+            type: 'del',
+            key: Buffer.from(key, 'hex'),
+          })
+        } else {
+          batchOp.push({
+            type: 'put',
+            key: Buffer.from(key, 'hex'),
+            value,
+          })
+        }
+      })
+      await this.batch(batchOp)
+    } else {
+      // dump everything into the current (higher level) cache
+      const currentKeyValueMap = this.checkpoints[this.checkpoints.length - 1].keyValueMap
+      keyValueMap.forEach(function (value, key) {
+        currentKeyValueMap.set(key, value)
+      })
```

Comment:
more of a nit than anything else, but this could be written shorthand in ES6:
`keyValueMap.forEach((value, key) => currentKeyValueMap.set(key, value))`